### PR TITLE
Rewrite state machine, add new state foundation

### LINKED ISSUE #36: Graceful Program Exit
# Graceful Program Exit

## Problem

When CTRL+C is used to end the program, the drone should be put into "Return" mode before the program ends.

## Solution

Catch CTRL+C exception and engage "Return" flight mode before ending program.

### LINKED ISSUE #37: Detect Kill-Switch
# Detect Kill-Switch

## Problem

Detect when the kill-switch is engaged by the remote. This occurs during a manual takeover. The program should end after the kill-switch is engaged to prevent additional commands from being sent to the drone.

## Solution

Detect when the kill-switch is engaged using `pymavlink`. When it is engaged, exit the program.

### LINKED ISSUE #128: Rewrite State Machine
# State Machine Revamp

## Problem

The current state machine has a lot of problems with circular imports and inconsistent states, which makes it currently impossible to do a full run through of all the states and features. 

## Solution

The structure of the state machine and all of the states should be planned out and implemented in order to avoid circular imports. Adding the ability to pass a list of states to be run in the state machines run function could also be useful for creating unit tests for singular states.

## Additional Information

Branch is feature/new_state_machine

### PR DESCRIPTION
**Pull request by Ooglely**
Replacing the old, broken state machine and flight manager with a new one using multiprocessing and state constructors in order to create a state machine that avoids the problems of circular imports.

### CODE REVIEW

File: state_machine/flight_manager.py [hunk 1]
```diff
@@ -0,0 +1,106 @@
...
+        except KeyboardInterrupt:
+            logging.critical(
+                "Keyboard interrupt detected. Killing state machine and landing drone."
+            )
+        finally:
+            state_machine.terminate()
+            asyncio.run(self.graceful_exit(drone_obj))
+
+    def start_state_machine(self, drone: Drone) -> None:
+        """Create and run a state machine in the event loop."""
```
Comment (cjhr95) [inline state_machine/flight_manager.py]: Missing function docstring

File: state_machine/flight_manager.py [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+        logging.info("-- Starting state machine")
+        asyncio.run(StateMachine(Start(drone), drone).run())
+
+    def start_kill_switch(self, process: Process, drone: Drone) -> None:
+        """Create and run a kill switch in the event loop."""
```
Comment (cjhr95) [inline state_machine/flight_manager.py]: Also missing full docstring

File: state_machine/flight_manager.py [hunk 3]
```diff
@@ -0,0 +1,106 @@
...
+        logging.info("-- Starting kill switch")
+        asyncio.run(self.kill_switch(process, drone))
+
+    async def kill_switch(self, state_machine_process: Process, drone: Drone) -> None:
+        """
+        Continuously check for whether or not the kill switch has been activated.
```
Comment (cjhr95) [inline state_machine/flight_manager.py]: Hrrm, interesting. Also missing a function docstring

File: state_machine/flight_manager.py [hunk 4]
```diff
@@ -0,0 +1,106 @@
...
+        async for flight_mode in drone.system.telemetry.flight_mode():
+            while flight_mode != FlightMode.MANUAL:
+                time.sleep(1)
+
+        logging.critical("Kill switch activated. Terminating state machine.")
+        state_machine_process.terminate()
+        return
+
+    async def graceful_exit(self, drone: Drone) -> None:
+        """Land the drone and exit the program."""
```
Comment (cjhr95) [inline state_machine/flight_manager.py]: Another one. Missing function docstring

File: flight/test_files/state_machine_test.py [hunk 5]
```diff
@@ -0,0 +1,20 @@
...
+
+SIM_ADDR: str = "udp://:14540"  # Address to connect to the simulator
+CONTROLLER_ADDR: str = "serial:///dev/ttyUSB0"  # Address to connect to a pixhawk board
+
+
+async def run_test(_sim: bool) -> None:  # Temporary fix for unused variable
+    """Runs the state machine.
+
+    Args:
+        _sim (bool): Whether or not to run the state machine in simulation mode.
```
Comment (cjhr95) [inline flight/test_files/state_machine_test.py]: The sim parameter isn't used in this function. In the flight_manager file, the start_manager() function has no way to distinguish between a simulated flight or a real one.

File: flight/test_files/state_machine_test.py [hunk 6]
```diff
@@ -0,0 +1,20 @@
+"""File that runs the state machine and kill switch in separate processes in order to test them."""
+import asyncio
+from state_machine.flight_manager import FlightManager
+
+SIM_ADDR: str = "udp://:14540"  # Address to connect to the simulator
+CONTROLLER_ADDR: str = "serial:///dev/ttyUSB0"  # Address to connect to a pixhawk board
+
+
+async def run_test(_sim: bool) -> None:  # Temporary fix for unused variable
+    """Runs the state machine.
```
Comment (cjhr95) [inline flight/test_files/state_machine_test.py]:
```suggestion
    """
    Runs the state machine.
```

File: flight/test_files/state_machine_test.py [hunk 7]
```diff
@@ -0,0 +1,20 @@
...
+
+    Args:
```
Comment (cjhr95) [inline flight/test_files/state_machine_test.py]:
```suggestion
    Parameters
    ----------
```

File: state_machine/drone.py [hunk 8]
```diff
@@ -0,0 +1,20 @@
...
+
+class Drone:
+    """
+    A drone for the state machine to control.
+    This class is a wrapper around the mavsdk System class, and will be passed around to each state.
+    Data can be stored in this class to be shared between states.
+    """
+
+    def __init__(self, address: str = "udp://:14540") -> None:
+        """Initialize a new Drone object."""
```
Comment (cjhr95) [inline state_machine/drone.py]: Needs function docstring

File: state_machine/drone.py [hunk 9]
```diff
@@ -0,0 +1,20 @@
+"""Defines the Drone class for the state machine."""
+
+import mavsdk
+
+
+class Drone:
+    """
+    A drone for the state machine to control.
+    This class is a wrapper around the mavsdk System class, and will be passed around to each state.
+    Data can be stored in this class to be shared between states.
```
Comment (cjhr95) [inline state_machine/drone.py:10]: Class docstring needs section for class variables and class methods.

File: state_machine/flight_manager.py [hunk 10]
```diff
@@ -0,0 +1,106 @@
...
+
+import asyncio
+import logging
+from multiprocessing import Process
+import time
+from mavsdk.telemetry import FlightMode, LandedState
+
+from .drone import Drone
+from .state_machine import StateMachine
+from .states import Start
```
Comment (cjhr95) [inline state_machine/flight_manager.py]: We usually stick to absolute imports, but if changing this yields some wacky errors, let me know.

File: state_machine/flight_manager.py [hunk 11]
```diff
@@ -0,0 +1,106 @@
...
+        Continuously checks for whether or not the kill switch has been activated.
+        Kills the state machine if the kill switch has been activated.
+    graceful_exit(drone: Drone)
+        Lands the drone and exits the program.
+    """
+
+    def __init__(self) -> None:
+        pass
+
+    def start_manager(self) -> None:
```
Comment (cjhr95) [inline state_machine/flight_manager.py:39]: There's no way for this function to initialize a drone with the flight controller's address versus the simulator currently. The sim parameter in the testing files does not make it here.

File: state_machine/states/impl/land_impl.py [hunk 12]
```diff
@@ -0,0 +1,51 @@
+import asyncio
+import logging
+
+from ..land import Land
+from ..start import Start
+from ..state import State
```
Comment (cjhr95) [inline state_machine/states/impl/land_impl.py]: Again, with the imports, if absolute imports destroy the code let me know.

File: state_machine/states/impl/start_impl.py [hunk 13]
```diff
@@ -0,0 +1,68 @@
+import asyncio
+import logging
+
+from ..start import Start
+from ..state import State
+from ..takeoff import Takeoff
```
Comment (cjhr95) [inline state_machine/states/impl/start_impl.py]: Relative imports? In my code? More likely than you think

File: state_machine/states/impl/takeoff_impl.py [hunk 14]
```diff
@@ -0,0 +1,50 @@
+import asyncio
+import logging
+
+from ..state import State
+from ..takeoff import Takeoff
+from ..waypoint import Waypoint
```
Comment (cjhr95) [inline state_machine/states/impl/takeoff_impl.py]: I think you get the jist

File: state_machine/states/impl/waypoint_impl.py [hunk 15]
```diff
@@ -0,0 +1,54 @@
+import asyncio
+import logging
+
+from ..land import Land
+from ..state import State
+from ..waypoint import Waypoint
```
Comment (cjhr95) [inline state_machine/states/impl/waypoint_impl.py]: More relative imports? You guys love these things

File: state_machine/states/land.py [hunk 16]
```diff
@@ -0,0 +1,41 @@
...
+from typing import Awaitable, Callable, ClassVar
+
+from .state import State
+
+
+class Land(State):
+    """
+    The Land state of the state machine.
+
+    Attributes:
```
Comment (cjhr95) [inline state_machine/states/land.py]:
```suggestion
    Attributes
    ----------
```

File: state_machine/states/land.py [hunk 17]
```diff
@@ -0,0 +1,41 @@
...
+        run_callable (ClassVar[Callable[["Land"], Awaitable[State]]]):
+            Class-level variable to hold a callable function signature.
+            It's a Callable that takes an instance of Land and returns an Awaitable[State].
+
+    Methods:
```
Comment (cjhr95) [inline state_machine/states/land.py]:
```suggestion
    Methods
    -------
```

File: state_machine/states/land.py [hunk 18]
```diff
@@ -0,0 +1,41 @@
...
+        run() -> Awaitable[State]:
+            Execute the logic associated with the Land state.
+
+            This method is called to perform the logic associated with the Land state.
+            It should return an Awaitable[State], which represents the next state
+            to transition to in the state machine.
+
+            Returns:
```
Comment (cjhr95) [inline state_machine/states/land.py]:
```suggestion
      Returns
      -------
```
Decrease the indent so it matches the other headers

File: state_machine/states/start.py [hunk 19]
```diff
@@ -0,0 +1,38 @@
...
+    run_callable : Callable[["Start"], Awaitable[State]]
+        The callable object which gets executed when the `run` method is called.
+
+    Methods
+    -------
+    run() -> Awaitable[State]
+        Executes the start state using the defined callable and transitions to the next state.
+
+        Returns
+        -------
```
Comment (cjhr95) [inline state_machine/states/start.py]: Decrease the indentation on this guy as well to match the other headers

File: state_machine/states/state.py [hunk 20]
```diff
@@ -0,0 +1,81 @@
+from abc import ABC, abstractmethod
+import logging
+from typing import Awaitable
+
+from ..drone import Drone
```
Comment (cjhr95) [inline state_machine/states/state.py]: Relative import yet again

**Review by cjhr95 — CHANGES_REQUESTED** [review]
Lots of missing function docstrings

**Comment by RexBerry** [comment]
I'm not sure switching to absolute imports would be easy given that there some imports from a parent package.

**Review by cjhr95 — APPROVED** [review]

**Pull request merged**